blocks/library/categories/index.js

```diff
@@ -0,0 +1,241 @@
+						!! childCategories.length && (
+							<ul className={ this.getCategoryListClassName( level + 1 ) }>
+								{ childCategories.map( childCategory => this.renderCategoryListItem( childCategory, level + 1 ) ) }
+							</ul>
+						)
+					}
+				</li>
+			);
+		}
+
+		renderCategoryDropdown() {
+			const { showHierarchy } = this.props.attributes;
+			const parentId = showHierarchy ? 0 : null;
+			const categories = this.getCategories( parentId );
+
+			return (
+				<select className={ `${ this.props.className }__dropdown` }>
+					{ categories.map( category => this.renderCategoryDropdownItem( category, 0 ) ) }
+				</select>
+			);
+		}
+
+		renderCategoryDropdownItem( category, level ) {
+			const { showHierarchy, showPostCounts } = this.props.attributes;
+			const childCategories = this.getCategories( category.id );
+
+			return [
+				<option key={ category.id }>
+					{ new Array( level * 3 ).fill( '\xa0' ) }
+					{ category.name.trim() || __( '(Untitled)' ) }
+					{
+						!! showPostCounts
+							? ` ( ${ category.count } )`
+							: ''
+					}
+				</option>,
+				showHierarchy &&
+				!! childCategories.length && (
+					childCategories.map( childCategory => this.renderCategoryDropdownItem( childCategory, level + 1 ) )
+				),
+			];
+		}
+
+		render() {
+			const { setAttributes } = this.props;
+			const categories = this.getCategories();
+
+			if ( ! categories.length ) {
+				return (
+					<Placeholder
+						icon="admin-post"
+						label={ __( 'Categories' ) }
+					>
+						<Spinner />
+					</Placeholder>
+				);
+			}
+
+			const { focus } = this.props;
+			const { align, displayAsDropdown, showHierarchy, showPostCounts } = this.props.attributes;
+
+			return [
+				focus && (
+					<BlockControls key="controls">
+						<BlockAlignmentToolbar
+							value={ align }
+							onChange={ ( nextAlign ) => {
+								setAttributes( { align: nextAlign } );
+							} }
+							controls={ [ 'left', 'center', 'right', 'full' ] }
+						/>
+					</BlockControls>
+				),
+				focus && (
+					<InspectorControls key="inspector">
+						<BlockDescription>
+							<p>{ __( 'Shows a list of your site\'s categories.' ) }</p>
+						</BlockDescription>
+						<h3>{ __( 'Categories Settings' ) }</h3>
+						<ToggleControl
+							label={ __( 'Display as dropdown' ) }
+							checked={ displayAsDropdown }
+							onChange={ this.toggleDisplayAsDropdown }
+						/>
+						<ToggleControl
+							label={ __( 'Show post counts' ) }
+							checked={ showPostCounts }
+							onChange={ this.toggleShowPostCounts }
+						/>
+						<ToggleControl
+							label={ __( 'Show hierarchy' ) }
+							checked={ showHierarchy }
+							onChange={ this.toggleShowHierarchy }
+						/>
+					</InspectorControls>
+				),
+				<div key="categories" className={ this.props.className }>
+					{
+						displayAsDropdown
+							? this.renderCategoryDropdown()
+							: this.renderCategoryList()
+					}
+				</div>,
+			];
+		}
+
+		componentWillUnmount() {
```

Comment:
Not prescribed, but general convention that `render` should be last method defined in a component class. I tend to prefer to keep lifecycle methods together toward the top of the class.

What do you think? Maybe we should document this, since I find I'm recommending it fairly often.